active/0012-cluster-call.md

```diff
@@ -0,0 +1,179 @@
+## Design
+
+This proposal uses mnesia to record the execution status of MFA, to ensure the consistency of the final resources & data in the cluster.
+
+This proposal is not applicable to high frequency request calls, all updates are performed in strict order.
+
+### mnesia table structure
+
+```erlang
+-record(cluster_rpc, {tnx_id :: pos_integer(), mfa :: mfa()}).
+-record(cluster_tnx_id, {node :: node(), tnx_id :: intger()}).
+
+mnesia(boot) ->
+    ok = ekka_mnesia:create_table(emqx_cluster_rpc, [
+        {type, ordered_set},
+        {disc_copies, [node()]},
+        {local_content, true},
+        {record_name, activated_alarm},
+        {attributes, record_info(fields, cluster_rpc)}]),
+    ok = ekka_mnesia:create_table(emqx_cluster_tnx_id, [
+        {type, set},
+        {disc_copies, [node()]},
+        {local_content, true},
+        {record_name, cluster_tnx_id},
+        {attributes, record_info(fields, cluster_tnx_id)}]);
+```
+
+`tnx_id` is strictly +1 incremental, all executed transactions must be executed in strict order, if there is node 1 executing transaction 1, 2, 3, but node 2 keeps failing in executing transaction 2 after executing transaction 1, it will keep retrying transaction 2 until it succeeds before executing transaction 3.
+
+### Interaction flow
+
+![cluster-call](./0012-assets/cluster-call.png)
+
+1. `emqx_cluster_trans` register on each node, subscribes to the mnesia table simple event, and is responsible for the execution of all transactions.
+2. The first transaction must be executed in the `emqx_cluster_trans` process on the core node. if this transaction succeeds, the call returns success directly, if the transaction fails, the call aborts with failure.
+
+```erlang
+    Node = node(),
+    case apply_self_unfinished_mfa_trans(Node) of
+        {error, Reason} -> {error, Reason};
+        ok ->
+            Trans = fun() ->
+                mnesia:write_lock_table(emqx_cluster_rpc),
+                mnesia:write_lock_table(emqx_cluster_tnx_id),
+                FinishId = get_finished_tnx_id(Node),
+                LatestId = get_latest_tnx_id(),
+                %% make sure node's local mfa all finished again.
+                case apply_self_unfinished_mfa(FinishId, LatestId, Node) of
+                    {error, Reason} -> mnesia:abort({error, Reason});
+                    ok ->
+                        TnxId = incr_tnx_id(),
+                        mnesia:write(#cluster_rpc{tnx_id = TnxId, mfa = MFA}),
+                        %% check new node join in but execute none transaction yet.
+                        [begin
+                             case mnesia:read(emqx_cluster_tnx_id, N) of
+                                 [] -> mnesia:write(#cluster_tnx_id{node = N, tnx_id = TnxId - 1});
+                                 _ -> ok
+                             end
+                         end || N <- ekka:nodelist()],
+                        mnesia:write(#cluster_tnx_id{node = Node, tnx_id = TnxId}),
```

Comment:
my thoughts about the usage of `emqx_cluster_tnx_id`:
This table is used to keep track of the last applied transaction id of their OWN.
The transaction initiator should NOT write records for other nodes.

So there is no need to lock `emqx_cluster_tnx_id` table in the beginning of the transaction.